akka-remote/src/main/scala/akka/remote/serialization/ArteryControlMessageSerializer.scala

```diff
@@ -0,0 +1,200 @@
+  }).toByteArray
+
+  override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = manifest match {
+    case QuarantinedManifest ⇒ deserializeQuarantined(ArteryControlFormats.Quarantined.parseFrom(bytes))
+    case ActorSystemTerminatingManifest ⇒ deserializeWithFromAddress(bytes, ActorSystemTerminating)
+    case ActorSystemTerminatingAckManifest ⇒ deserializeWithFromAddress(bytes, ActorSystemTerminatingAck)
+    case HandshakeReqManifest ⇒ deserializeWithFromAddress(bytes, HandshakeReq)
+    case HandshakeRspManifest ⇒ deserializeWithFromAddress(bytes, HandshakeRsp)
+    case ActorRefCompressionAdvertisementManifest ⇒ deserializeCompressionAdvertisement(bytes, deserializeActorRef, ActorRefCompressionAdvertisement)
+    case ActorRefCompressionAdvertisementAckManifest ⇒ deserializeCompressionTableAdvertisementAck(bytes, ActorRefCompressionAdvertisementAck)
+    case ClassManifestCompressionAdvertisementManifest ⇒ deserializeCompressionAdvertisement(bytes, identity, ClassManifestCompressionAdvertisement)
+    case ClassManifestCompressionAdvertisementAckManifest ⇒ deserializeCompressionTableAdvertisementAck(bytes, ClassManifestCompressionAdvertisementAck)
+    case SystemMessageEnvelopeManifest ⇒ deserializeSystemMessageEnvelope(bytes)
+    case SystemMessageDeliveryAckManifest ⇒ deserializeSystemMessageDeliveryAck(bytes, SystemMessageDelivery.Ack)
+    case SystemMessageDeliveryNackManifest ⇒ deserializeSystemMessageDeliveryAck(bytes, SystemMessageDelivery.Nack)
+    case _ ⇒ throw new IllegalArgumentException(s"Manifest '$manifest' not defined for ArteryControlMessageSerializer (serializer id $identifier)")
+  }
+
+  import scala.collection.JavaConverters._
+
+  def serializeQuarantined(quarantined: Quarantined): ArteryControlFormats.Quarantined =
+    ArteryControlFormats.Quarantined.newBuilder()
+      .setFrom(serializeUniqueAddress(quarantined.from))
+      .setTo(serializeUniqueAddress(quarantined.to))
+      .build
+
+  def deserializeQuarantined(quarantined: ArteryControlFormats.Quarantined): Quarantined =
+    Quarantined(deserializeUniqueAddress(quarantined.getFrom), deserializeUniqueAddress(quarantined.getTo))
+
+  def serializeActorRef(ref: ActorRef): String =
+    if (ref == system.deadLetters) "DEADLETTER"
+    else Serialization.serializedActorPath(ref)
+
+  def deserializeActorRef(str: String): ActorRef =
+    if (str == "DEADLETTER") system.deadLetters
+    else system.provider.resolveActorRef(str)
+
+  def serializeCompressionAdvertisement[T](adv: CompressionAdvertisement[T])(keySerializer: T ⇒ String): ArteryControlFormats.CompressionTableAdvertisement = {
+    val builder =
+      ArteryControlFormats.CompressionTableAdvertisement.newBuilder
+        .setFrom(serializeUniqueAddress(adv.from))
+        .setTableVersion(adv.table.version)
+
+    adv.table.map.foreach {
+      case (key, value) ⇒
+        builder
```

Comment:
... instead look for value == 0 here

but we should look at how 0 is used, and why it's in the advertised table in the first place. It's well know on both sides, so should not be needed.